internal/providers/manager/manager.go

```diff
@@ -104,9 +111,84 @@ func NewProviderManager(
 	return &providerManager{
 		classManagers: classes,
 		store:         store,
+		sessionStore:  sessionStore,
 	}, nil
 }
 
+func (p *providerManager) CreateFromConfig(ctx context.Context, providerClass db.ProviderClass, projectID uuid.UUID, name string, config json.RawMessage) (*db.Provider, error) {
+	manager, err := p.getClassManager(providerClass)
+	if err != nil {
+		return nil, fmt.Errorf("error getting class manager: %w", err)
+	}
+
+	provConfig, err := manager.GetConfig(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting provider config: %w", err)
+	}
+
+	// This method feels like it should be part of the class manager?
+	providerDef, err := providers.GetProviderClassDefinition(string(providerClass))
+	if err != nil {
+		return nil, fmt.Errorf("error getting provider definition: %w", err)
+	}
+
+	providerParams := db.CreateProviderParams{
+		Name:       name,
+		ProjectID:  projectID,
+		Class:      providerClass,
+		Implements: providerDef.Traits,
+		Definition: provConfig,
+		AuthFlows:  providerDef.AuthorizationFlows,
+	}
+
+	return p.store.Create(ctx, projectID, &providerParams)
+}
+
+func (p *providerManager) CreateFromSessionState(ctx context.Context, providerClass db.ProviderClass, creds v1.Credential, state string) (*db.Provider, error) {
```

Comment:
As discussed earlier: I don't think `CreateFromSessionState` belongs in `ProviderManager`. The session state is one mechanism by which providers can be created, and I don't think the `ProviderManager` should need to know or care about that some providers get created this way.

I would factor this session state code into a `SessionStateService` or similar, and let that call the `CreateFromConfig` method on provider manager, but take care of the other logic itself.